secure_delete/sswap.1
.\" This definition swiped from the gcc(1) man page
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH SSWAP 1

.SH NAME
sswap \- secure swap wiper (secure_deletion toolkit)

.SH SYNOPSIS
.B sswap [-f] [-l] [-l] [-v] [-z] swapdevice

.SH DESCRIPTION
.I sswap
is designed to delete data which may lie still on your swapspace
in a secure manner which can not be recovered by thiefs, law enforcement
or other threats.
The wipe algorythm is based on the paper "Secure Deletion of Data from
Magnetic and Solid-State Memory" presented at the 6th Usenix Security
Symposium by Peter Gutmann, one of the leading civilian cryptographers.
.PP
The
.I secure data deletion
process of sswap goes like this:
.PP
.TP
.B *
1 pass with 0xff
.TP
.B * 
5 random passes. /dev/urandom is used for a secure RNG if available.
.TP
.B * 
27 passes with special values defined by Peter Gutmann.
.TP
.B * 
5 random passes. /dev/urandom is used for a secure RNG if available.
.PP

.SH COMMANDLINE OPTIONS
.PP
.TP
.B \-f
fast (and insecure mode): no /dev/urandom, no synchronize mode.
.TP
.B \-l 
lessens the security. Only two passes are written: one mode with 0xff and
a final mode with random values.
.TP
.B \-l
-l for a second time lessons the security even more: only one pass with
random values is written.
.TP
.B \-v
verbose mode
.TP
.B \-z
wipes the last write with zeros instead of random data
.PP

.SH BEWARE
.TP
.B swapoff
unmount your swapspace before using this tool! Otherwise your system might
crash!
.TP
.B BETA!
.I sswap
is still beta. It was only tested on Linux but on this system it performed
it's work all of the time.

.PP
.SH BUGS 
No bugs. There was never a bug in the secure_deletion package (in contrast
to my other tools, whew, good luck ;-)
Send me any that you find.  Patches are nice too :)

.SH AUTHOR
.Sp
van Hauser / THC
.I <vh@thc.org>

.SH DISTRIBUTION
The newest version of the
.I secure_deletion package
can be obtained from 
.I http://www.thc.org
.Sp
.I sswap
and the
.I secure_deletion package
is (C) 1997-2003 by van Hauser / THC (vh@thc.org)
.Sp
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; Version 2.
.Sp
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

.SH SEE ALSO
.I srm
(1),
.I sfill
(1),
.I smem
(1)


secure_delete/CHANGES
v3.1 (November 2003)	PUBLIC RELEASE
	(sdel-mod)	I added a kernel module in v3.0 which wipes all
			files which are unlinked() in the kernel. However
			it didnt work, so it was not official.
			However, deathcubek@md2600.net sent in a patch,
			so now it is working :-)
			After "make install" do "insmod sdel-mod"
	(srm)		Fixed a bug which prevented to delete directories

v3.0 (April 2003)	PUBLIC RELEASE
	(all)		Added >2GB file support
	(all)		rewrote all tools to use a central library, this
			was long time necessary!
	(all)		warning now displayed if ulimit value resetting
			support was not compiled in.
	(sswap)		added -j(ump) option to skip over a defined number
			of bytes of the block device.
	(srm)		skipping file wipe if it is hardlinked!
	(fileutils-rm)	Update the rm.diff for the newest fileutil (4.1)
	(all)		updated the link for our homepage + email to thc.org

v2.5 (May 2001)		PUBLIC RELEASE
	(all)		updated the link for our homepage to
			www.thehackerschoice.com

v2.4 (October 2000)	PRIVATE RELEASE
	(all)		added the option -z to write zeros as last wipe
			instead of random data
	(sfill)		added the options -i and -I to not or exclusivly wipe
                        inode space and expanded the number of inodes to wipe
	(the_cleaner.sh) added a new shellscript called the_cleaner.sh which
			wipes temporary directories, all user's history and
			netscape history/cache files, free disk and inode
			space, swap space and the memory. Easy shellscript
			to place in the shutdown phase of runlevel 2 and 3!
	(all)		beautified the help output; updated documentation
	(all)		added srm/sfill/sswap/smem as linux elf binaries

v2.3 (August 2000)	PUBLIC RELEASE
	(sfill)		now also wipes all inodes in the directory specified
	(srm)		fixed a bug in srm when deleting dirs without -r option
			(thanks to typo/teso)

v2.2 (October 1999)	PUBLIC RELEASE
	(WWW)		ugh, the redirector service base.org is down!
			WWW site: http://www.infowar.co.uk/thc
	(all)		code cleanup to prevent warnings (thanks Rebecca!)
			(some weird systems need O_FSYNC and strings.h)
	(srm)		the blocksize used for wiping is now enlarged if the
			filesystem is detected to use bigger blocks.
	(srm/sfill/sswap) changed the behaviour, now the last wipe is always
			a random one, so it's harder to find out that a
			secure deletion was done. (Thanks to Dan Farmer for
			this idea)
	(srm)		the random buffer to write was not renewed after one
			write to the file, this makes it pretty fast but also
			not that secure. No all random writes are unique. The
			other tools had that from the beginning on. duh

v2.1 (April 1999)	PRIVATE RELEASE
	(all)		added the -f(ast) option.
	(smem)		added a new tool: smem, which wipes free memory
	(srm/sfill/smem) added rlimit resets
	(sfill/sswap)	quadroupled the buffer size
	(srm/sfill/sswap) fix for openbsd which doesn't know O_SYNC open(2).
        (srm)		added the -d(ot) option. Ignores "." and ".." on the
			commandline
	(srm)		fixed directory and file link races.
			Thanks to Solar Designer bringing this to my attention.
			Note that there is a small racecondition left which
			can be exploited from an attacker to chmod 0600 a file
			to it's owner if the owner is running srm and the
			attacker has write access to the target directory.
			The other wipe programs have this problem too - and
			( - wiping anything in your tree for which you
			have write permission ... - ) worse.

v2.0
	(all)		added man pages for srm, sfill and sswap!
	(all)		increased speed by ~300% by writing big buffers.
	(all)		using /dev/urandom for random passes now if available
			(this is more secure, however costs speed)
	(all)		added the O_SYNC bit to all open calls. Should not
			add security, but who knows how weired systems behave.
	(all)		the -s option(s) is now ignored. The standard mode
			is now the secure mode. Use -l to lessen the security.
	(srm)		added the option -r for directory recursion
	(srm)		doing a chmod() now before opening the file
	(srm)		fixed handling of symlink and special files
	(rm.diff)	includes all the above except for the new -s/-l change
	(all)		cosmetics + documenation updates

v1.8
	(rm.diff)	updated the diff and did some more security and speed.
	(all)		a bit cosmetics

v1.7
	(srm)		using fsync() now, which gives you 10-25% speed !
	(all)		using perror() for better error reporting
	(srm/sfill)	included file truncating before unlinking (security!)

v1.6
	[www distrib.]	http://r3wt.base.org - check out the THC area
	[email]		my new email: vh@reptile.rug.ac.be
	(sswap)		added a new tool: a secure swap cleaner. remember to
			unmount your swap first. *only tested on linux* !
	(srm/sfill) 	added trapping signals and exiting clean on them
	(srm/sfill)	errors are sent do stderr now instead to stdout
	(srm/sfill)	general code cleanup
	(sfill)		fixed race condition to prevent symlink attacks
	(sfill)		warning if executed by user!=root, because not all
			free space can be overwritten (idea by fyodor@dhp.com)
	(rm.diff)	replaces the rm.c replacement with a diff (+ bugfix)
	(sdel.exe/sfill.exe) removed the msdos versions. They 1st already
			included the features above and 2nd are no longer
			supported (at least by me - get a good unix)

v1.4
	(sfill)		some flushing on stdout done
	(sfill)		more error checking
	(srm/sfill)	cosmetics

v1.3
			initial public release.


secure_delete/sswap.c
/* Secure SWAP cleaner - by van Hauser / [THC], vh@thc.org
 *
 * Note that this program is beta. It was tested with linux, but I can't
 * tell for other platforms. Read the statement at #define SWAP_PAGESIZE
 * on how to use this program on other unix machines.
 *
 * of course: you have to turn of the swapspace before using this program !
 *
 * Secure SWAP overwrites all data on your swap device.
 * Standard mode is a real security wipe for 38 times, flushing
 * the caches after every write. The wipe technique was proposed by Peter
 * Gutmann at Usenix '96 and includes 10 random overwrites plus 28 special
 * defined characters. Take a look at the paper of him, it's really worth
 * your time.
 * The option -l overwrites two times the data. (0xff + random)
 * The option -ll overwrites the data once. (random)
 *
 * Read the manual for limitations.
 * Compiles clean on OpenBSD, Linux, Solaris and AIX
 *
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <strings.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>

#include "sdel.h"

/* SWAP_PAGESIZE is an important variable. You have to set this
 * to your header length of your swapdevice. For Linux this is 4096,
 * I don't know for the other OSs. To be sure, set this to 0 and
 * recreate your swapspace afterwards (for linux: mkswap /dev/swapdevice)
 */
#define SWAP_PAGESIZE	4096
#ifdef BLOCKSIZE
 #undef BLOCKSIZE
#endif
#define BLOCKSIZE	65535

int fd;
int slow = O_SYNC;
int zero = 0;

void help() {
    printf("sswap %s (c) 1997-2003 by %s <%s>\n\n", VERSION, AUTHOR, EMAIL);
    printf("Syntax: %s [-flvz] [-j start] /dev/of_swap_device\n\n", prg);
    printf("Options:\n");
    printf("\t-f  fast (and insecure mode): no /dev/urandom, no synchronize mode.\n");
    printf("\t-j  jump over the first number of bytes when wiping. (default: %d)\n", SWAP_PAGESIZE);
    printf("\t-l  lessens the security (use twice for total insecure mode).\n");
    printf("\t-v  is verbose mode.\n");
    printf("\t-z  last wipe writes zeros instead of random data.\n");
    printf("\nsswap does a secure overwrite of the swap space.\n");
    printf("Default is secure mode (38 writes).\n");
    printf("Updates can be found at %s\n", WEB);
    printf("\nNOTE: You must disable the swapspace before using this program!\007\n");
    exit(1);
}

void cleanup() {
    fprintf(stderr,"\nTerminated by signal. Clean exit.\n");
    close(fd);
    sync();
    exit(1);
}

int main (int argc, char *argv[]) {
    int secure = 2;
    int result;
    int mode;
    unsigned long start = SWAP_PAGESIZE;
    struct stat stats;
    char *filename;
    
    prg = argv[0];
    if (argc == 1 || strncmp(argv[1], "-h", 2) == 0 || strncmp(argv[1], "--h", 3) == 0)
        help();

    while (1) {
        result = getopt(argc, argv, "FfJ:j:LlSsVvZz");
        if (result<0) break;
        switch (result) {
            case 'F' :
            case 'f' : slow = 0;
                       break;
            case 'J' :
            case 'j' : start = atol(optarg);
                       if (start < 0 || start > 65535) {
                           fprintf(stderr, "Error: The -j option must be set between 0 and 65535!\n");
                           exit(-1);
                       }
                       break;
            case 'L' :
            case 'l' : if (secure) secure--;
            	       break;
            case 'S' :
            case 's' : secure++;
                       break;
            case 'V' :
            case 'v' : verbose++;
                       break;
            case 'Z' :
            case 'z' : zero++;
                       break;
            default :  help();
        }
    }

    if ((optind+1) != argc)
        help();

    signal(SIGINT, cleanup);
    signal(SIGTERM, cleanup);
    signal(SIGHUP, cleanup);

    filename = argv[optind];
    mode = secure;

    if ((fd = open (filename, O_RDWR | O_LARGEFILE | slow)) < 0) {
        fprintf(stderr, "Error: Can't open %s for writing.", filename);
        perror("");
        exit(1);
    }

    fstat(fd, &stats);
    if (!S_ISBLK(stats.st_mode)) {
        fprintf(stderr, "Error: Target is not a block device - %s\n", filename);
        exit(1);
    }

    if (verbose) {
        char type[15] = "random";
        if (zero) strcpy(type, "zero");
        switch (mode) {
           case 0 : printf("Wipe mode is insecure (one pass [%s])\n",type);
                    break;
           case 1 : printf("Wipe mode is insecure (two passes [0xff/%s])\n",type);
                    break;
           default: printf("Wipe mode is secure (38 special passes)\n");
        }
 	printf("Writing to device %s: ", filename);
    }

    sdel_init(slow);

    if (sdel_overwrite(mode, fd, start, BLOCKSIZE, 0, zero) == 0)
        if (verbose)
    	    printf(" done\n");

    sdel_finnish();

    /* thats all */
    exit(0);
}
